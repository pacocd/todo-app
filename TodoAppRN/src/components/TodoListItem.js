import React, { Component } from 'react';
import { Text, View, Switch } from 'react-native';
import PropTypes from 'prop-types';

export class TodoListItem extends Component {
  static propTypes = {
    todo: PropTypes.shape({
      id: PropTypes.number.isRequired,
      user_id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      completed: PropTypes.bool.isRequired
    }).isRequired
  };

  render() {
    const { title } = this.props.todo;
    return (
      <View style={{ width: '100%', paddingVertical: 14 }}>
        <Text>{title}</Text>
      </View>
    );
  }
}

export default TodoListItem;
