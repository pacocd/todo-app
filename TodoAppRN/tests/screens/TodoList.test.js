import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {
  TodoList,
  mapDispatchToProps,
  mapStateToProps
} from '../../src/screens/TodoList';
import TodoListItem from '../../src/components/TodoListItem';

configure({ adapter: new Adapter() });

describe('when redux state functions are called', () => {
  it('calls mapStateToProps to get todoListData from todos', () => {
    expect(mapStateToProps({ todos: { todoListData: [] } })).toEqual({
      todoListData: []
    });
  });

  it('calls mapDispatchToProps to get a new object', () => {
    const dispatchMock = jest.fn();
    expect(mapDispatchToProps(dispatchMock).setTodoListData).toBeInstanceOf(
      Function
    );
  });
});

describe('when TodoList is fully rendered', () => {
  const todoListDataMock = [
    { id: 1, title: 'test', user_id: 1, completed: false }
  ];
  const props = {
    navigation: {},
    getTodoList: jest.fn().mockReturnValue(
      new Promise(resolve =>
        resolve({
          data: todoListDataMock,
          status: 200
        })
      )
    ),
    hasError: jest.fn().mockReturnValue(false),
    setTodoListData: jest.fn(),
    todoListData: todoListDataMock
  };
  const wrapper = shallow(<TodoList {...props} />);

  beforeAll(async () => {
    await wrapper.instance().props.getTodoList();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('expects getTodoList prop function to have been called', () => {
    expect(wrapper.instance().props.getTodoList).toHaveBeenCalled();
  });

  it('returns an error in getTodoList response', () => {
    expect(wrapper.instance().props.hasError).toHaveBeenCalledWith({
      status: 200,
      data: todoListDataMock
    });
  });

  it('expects setTodoListData to have been called with data from getTodoList', () => {
    expect(wrapper.instance().props.setTodoListData).toHaveBeenCalledWith(
      todoListDataMock
    );
  });

  it('calls keyExtractor and expect to extract id of item as String', () => {
    const item = { id: 1 };
    expect(wrapper.instance().keyExtractor(item)).toBe('1');
  });

  it('calls renderListItem to get TodoListItem', () => {
    const itemData = {
      item: { id: 1, title: 'test', user_id: 1, completed: false }
    };
    expect(wrapper.instance().renderListItem(itemData)).toEqual(
      <TodoListItem todo={itemData.item} />
    );
  });

  it('expects todoFlatList to be rendered', () => {
    expect(
      wrapper
        .find('.todoListView')
        .find('.todoFlatList')
        .exists()
    ).toBeTruthy();
  });

  it('expects activityIndicatorView to not be rendered', () => {
    expect(wrapper.find('.activityIndicatorView').exists()).toBeFalsy();
  });
});

describe('when TodoList is loading', () => {
  const todoListDataMock = [
    { id: 1, title: 'test', user_id: 1, completed: false }
  ];
  const props = {
    navigation: {},
    getTodoList: jest.fn(),
    hasError: jest.fn().mockReturnValue(true),
    setTodoListData: jest.fn(),
    todoListData: undefined
  };
  const wrapper = shallow(<TodoList {...props} />);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('expects activityIndicatorView to be rendered', () => {
    expect(wrapper.find('.activityIndicatorView').exists()).toBeTruthy();
  });

  it('expects todoFlatList to not be rendered', () => {
    expect(
      wrapper
        .find('.todoListView')
        .find('.todoFlatList')
        .exists()
    ).toBeFalsy();
  });
});
