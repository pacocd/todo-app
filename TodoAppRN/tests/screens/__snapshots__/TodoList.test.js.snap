// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when TodoList is fully rendered renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    deleteTodo={[MockFunction]}
    getTodoList={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {
              "_40": 0,
              "_55": Object {
                "data": Array [
                  Object {
                    "completed": false,
                    "id": 1,
                    "title": "test",
                    "user_id": 1,
                  },
                ],
                "status": 200,
              },
              "_65": 1,
              "_72": null,
            },
          },
          Object {
            "isThrow": false,
            "value": Promise {
              "_40": 0,
              "_55": Object {
                "data": Array [
                  Object {
                    "completed": false,
                    "id": 1,
                    "title": "test",
                    "user_id": 1,
                  },
                ],
                "status": 200,
              },
              "_65": 1,
              "_72": null,
            },
          },
        ],
      }
    }
    hasError={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "data": Array [
                Object {
                  "completed": false,
                  "id": 1,
                  "title": "test",
                  "user_id": 1,
                },
              ],
              "status": 200,
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": false,
          },
        ],
      }
    }
    navigation={
      Object {
        "navigate": [MockFunction],
      }
    }
    setTodoListData={
      [MockFunction] {
        "calls": Array [
          Array [
            Array [
              Object {
                "completed": false,
                "id": 1,
                "title": "test",
                "user_id": 1,
              },
            ],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    todoListData={
      Array [
        Object {
          "completed": false,
          "id": 1,
          "title": "test",
          "user_id": 1,
        },
      ]
    }
    updateTodo={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <SwipeListView
        className="todoFlatList"
        closeOnRowBeginSwipe={false}
        closeOnRowPress={true}
        closeOnScroll={true}
        data={
          Array [
            Object {
              "completed": false,
              "id": 1,
              "title": "test",
              "user_id": 1,
            },
          ]
        }
        directionalDistanceChangeThreshold={2}
        disableLeftSwipe={false}
        disableRightSwipe={true}
        keyExtractor={[Function]}
        leftOpenValue={0}
        previewFirstRow={false}
        recalculateHiddenLayout={false}
        renderHiddenItem={[Function]}
        renderItem={[Function]}
        rightOpenValue={-75}
        swipeToClosePercent={50}
        swipeToOpenPercent={50}
        swipeToOpenVelocityContribution={0}
        useFlatList={true}
      />,
      "className": "todoListView",
      "style": Object {
        "backgroundColor": "white",
        "flex": 1,
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": "todoFlatList",
        "closeOnRowBeginSwipe": false,
        "closeOnRowPress": true,
        "closeOnScroll": true,
        "data": Array [
          Object {
            "completed": false,
            "id": 1,
            "title": "test",
            "user_id": 1,
          },
        ],
        "directionalDistanceChangeThreshold": 2,
        "disableLeftSwipe": false,
        "disableRightSwipe": true,
        "keyExtractor": [Function],
        "leftOpenValue": 0,
        "previewFirstRow": false,
        "recalculateHiddenLayout": false,
        "renderHiddenItem": [Function],
        "renderItem": [Function],
        "rightOpenValue": -75,
        "swipeToClosePercent": 50,
        "swipeToOpenPercent": 50,
        "swipeToOpenVelocityContribution": 0,
        "useFlatList": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <SwipeListView
          className="todoFlatList"
          closeOnRowBeginSwipe={false}
          closeOnRowPress={true}
          closeOnScroll={true}
          data={
            Array [
              Object {
                "completed": false,
                "id": 1,
                "title": "test",
                "user_id": 1,
              },
            ]
          }
          directionalDistanceChangeThreshold={2}
          disableLeftSwipe={false}
          disableRightSwipe={true}
          keyExtractor={[Function]}
          leftOpenValue={0}
          previewFirstRow={false}
          recalculateHiddenLayout={false}
          renderHiddenItem={[Function]}
          renderItem={[Function]}
          rightOpenValue={-75}
          swipeToClosePercent={50}
          swipeToOpenPercent={50}
          swipeToOpenVelocityContribution={0}
          useFlatList={true}
        />,
        "className": "todoListView",
        "style": Object {
          "backgroundColor": "white",
          "flex": 1,
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "todoFlatList",
          "closeOnRowBeginSwipe": false,
          "closeOnRowPress": true,
          "closeOnScroll": true,
          "data": Array [
            Object {
              "completed": false,
              "id": 1,
              "title": "test",
              "user_id": 1,
            },
          ],
          "directionalDistanceChangeThreshold": 2,
          "disableLeftSwipe": false,
          "disableRightSwipe": true,
          "keyExtractor": [Function],
          "leftOpenValue": 0,
          "previewFirstRow": false,
          "recalculateHiddenLayout": false,
          "renderHiddenItem": [Function],
          "renderItem": [Function],
          "rightOpenValue": -75,
          "swipeToClosePercent": 50,
          "swipeToOpenPercent": 50,
          "swipeToOpenVelocityContribution": 0,
          "useFlatList": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`when TodoList is loading renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    deleteTodo={[Function]}
    getTodoList={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    hasError={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": true,
          },
        ],
      }
    }
    navigation={Object {}}
    setTodoListData={[MockFunction]}
    updateTodo={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <ActivityIndicator
        animating={true}
        color="#999999"
        hidesWhenStopped={true}
        size="large"
      />,
      "className": "activityIndicatorView",
      "style": Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "flex": 1,
        "justifyContent": "center",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "animating": true,
        "color": "#999999",
        "hidesWhenStopped": true,
        "size": "large",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <ActivityIndicator
          animating={true}
          color="#999999"
          hidesWhenStopped={true}
          size="large"
        />,
        "className": "activityIndicatorView",
        "style": Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "justifyContent": "center",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "animating": true,
          "color": "#999999",
          "hidesWhenStopped": true,
          "size": "large",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
